### ISSUE TITLE
[Question] The source code for layer grad norm figures.

### ISSUE THREAD

**Issue by sherlcok314159** [BryceZhuo/HybridNorm#1]
Hello! You proposed an interesting norm type called HybridNorm. The layer grad norm figure is surprising. However, I did not find the source code for drawing this figure during the training. So, I wonder if I missed some code snippets or you forgot to add this part. :-)

Kind regards.

**Comment by BryceZhuo** [comment]
Hi!

Sorry for the late reply—I was traveling recently. Thanks for your interest in HybridNorm!

To record the gradient norm for each parameter during training, you can enable it by adding the following flag to your training script:
```
--optimizer.record_update_metrics=true \
--turn_off_optimizer_log=false
```
This will allow you to log the gradient statistics. The corresponding implementation can be found here:
https://github.com/BryceZhuo/HybridNorm/blob/f0dd60589c4cafcbd5496d60884d2745ff340805/dense_model/olmo/optim.py#L49C9-L49C39

Once this is set up, you can visualize the gradient norms using the data logged to Weights & Biases.

**Comment by sherlcok314159** [comment]
tks, I will check it out!

**Comment by sherlcok314159** [comment]
@BryceZhuo It seems like that I failed to reproduce your results on demo datasets. I pick two files from each training subset. Here is my config for the dense model. I only experimented with the dense model.

```yaml
  paths:
    # ProofPile 2: Algebraic Stack Data
    - http://olmo-data.org/preprocessed/proof-pile-2/v0_decontaminated/algebraic-stack/train/allenai/dolma2-tokenizer/part-01-00000.npy
    - http://olmo-data.org/preprocessed/proof-pile-2/v0_decontaminated/algebraic-stack/train/allenai/dolma2-tokenizer/part-02-00000.npy

    # ProofPile 2: Arxiv Data
    - http://olmo-data.org/preprocessed/proof-pile-2/v0_decontaminated/arxiv/train/allenai/dolma2-tokenizer/part-00-00000.npy
    - http://olmo-data.org/preprocessed/proof-pile-2/v0_decontaminated/arxiv/train/allenai/dolma2-tokenizer/part-01-00000.npy

    # ProofPile 2: Open Web Math Data
    - http://olmo-data.org/preprocessed/proof-pile-2/v0_decontaminated/open-web-math/train/allenai/dolma2-tokenizer/part-00-00000.npy
    - http://olmo-data.org/preprocessed/proof-pile-2/v0_decontaminated/open-web-math/train/allenai/dolma2-tokenizer/part-01-00000.npy

    # Pes2o Data
    - http://olmo-data.org/preprocessed/pes2o/allenai/dolma2-tokenizer/part-00-00000.npy
    - http://olmo-data.org/preprocessed/pes2o/allenai/dolma2-tokenizer/part-01-00000.npy

    # Starcoder Data (fixed!)
    - http://olmo-data.org/preprocessed/starcoder/v1-decon-100_to_20k-2star-top_token_030/allenai/dolma2-tokenizer/part-000-00000.npy
    - http://olmo-data.org/preprocessed/starcoder/v1-decon-100_to_20k-2star-top_token_030/allenai/dolma2-tokenizer/part-001-00000.npy

    # Wikipedia
    - http://olmo-data.org/preprocessed/olmo-mix/danyh-compiled-v1_7/documents/wiki/allenai/dolma2-tokenizer/part-0-00000.npy
    - http://olmo-data.org/preprocessed/olmo-mix/danyh-compiled-v1_7/documents/wiki/allenai/dolma2-tokenizer/part-1-00000.npy
```

The export to wandb code is as follows. I only ran the code for 1914 steps.

```python
import wandb
import pandas as pd

ENTITY = "xxx"
PROJECT = "hybridnorm"
RUN_ID = "etdem4d9"
SECTION_PREFIX = "optim/grad/"
PANEL_SUFFIX = "weight.norm"
OUTPUT_CSV_FILE = f"{RUN_ID}_optim_grad_data.csv"
PAGE_SIZE = 1E6
EXPECTED_ROWS_NUM = 1914

api = wandb.Api()
run_path = f"{ENTITY}/{PROJECT}/{RUN_ID}"
run = api.run(run_path)
sampled_history = run.history(pandas=True)
optim_grad_cols = [
    col for col in sampled_history.columns if col.startswith(SECTION_PREFIX) and
    col.endswith(PANEL_SUFFIX) and 'transformer.blocks.' in col
]
history = run.scan_history(page_size=PAGE_SIZE, keys=optim_grad_cols)
rows = [row for row in history]
assert len(
    rows
) == EXPECTED_ROWS_NUM, 'The number of rows is not equal to the one in the web interface.'
rows_df = pd.DataFrame(rows)
rows_df.to_csv(OUTPUT_CSV_FILE, index=False, encoding='utf-8')
```

The visualization code is as follows.

```python
import numpy as np
import os
import torch
import pandas as pd
import sys
import matplotlib.pyplot as plt


def visualize_layer_grad(iteration, layer_data):
    num_layers = len(layer_data)
    layer_indices = np.arange(num_layers)
    plt.style.use('seaborn-v0_8-whitegrid')
    bar_color = '#4285F4'
    fig_width = max(12, num_layers * 0.5)
    fig, ax = plt.subplots(figsize=(fig_width, 6))
    bars = ax.bar(
        layer_indices,
        layer_data,
        color=bar_color,
        alpha=0.85,
        edgecolor='black',
        linewidth=0.7
    )

    ax.bar_label(bars, padding=3, fmt='%.4f', fontsize=8)
    ax.set_xlabel("Layer Index", fontsize=12)
    ax.set_ylabel("Gradient L2 Norm", fontsize=12)
    ax.set_title(f"Gradient Norm per Layer at Iteration {iteration}", fontsize=14, pad=15)
    ax.set_xticks(layer_indices)
    ax.set_xticklabels(layer_indices, fontsize=10)
    ax.set_ylim(bottom=0, top=max(layer_data) * 1.15)
    ax.yaxis.grid(True, linestyle='--', which='major', color='grey', alpha=0.6)
    ax.xaxis.grid(False)
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)
    ax.spines['left'].set_linewidth(0.8)
    ax.spines['bottom'].set_linewidth(0.8)
    fig.tight_layout()
    os.makedirs('figs', exist_ok=True)
    plt.savefig(f'figs/{iteration}.png', dpi=400)


iteration = int(sys.argv[1])
data = pd.read_csv('etdem4d9_optim_grad_data.csv')
modules = data.columns.tolist()
iteration_values = data.iloc[iteration - 1].tolist()
layers = 16
layer_grad_norm = []
for layer_idx in range(layers):
    layer_weights_idxs = [
        idx for idx, name in enumerate(modules)
        if f'transformer.blocks.{layer_idx}' in name
    ]
    layer_weights_norms_squares = torch.tensor([
        iteration_values[idx]**2 for idx in layer_weights_idxs
    ])
    layer_grad_norm.append(layer_weights_norms_squares.sum().sqrt())

visualize_layer_grad(iteration, layer_grad_norm)
```

In the end, I chose many steps to reproduce your result. But all the results are quite different from yours in the paper.

![Image](https://github.com/user-attachments/assets/d1238630-61a0-40f0-95a0-31ad70ac34c7)

![Image](https://github.com/user-attachments/assets/9905a28a-e055-4e49-ae97-30234a461ef2)

![Image](https://github.com/user-attachments/assets/537ded3d-978d-4d09-91c8-f83c5d5bc84c)

**Comment by BryceZhuo** [comment]
Hi,
From the code perspective, the difference might be caused by the following two factors:
	1. Our gradient visualization mainly follows Figure 3 in [1], where only the gradients of the FFN weights are included in the statistics.
	2. The figure we plotted shows HybridNorm, not HybridNorm*. The first layer of HybridNorm* has some special handling, which can lead to slightly different gradients in the first one or two layers.
You could try adjusting based on the points above. If you still run into issues, feel free to post your results—we’d be happy to take another look.

[1] Xiong, et al. [On Layer Normalization in the Transformer Architecture](https://proceedings.mlr.press/v119/xiong20b/xiong20b.pdf). ICML 2020.

**Comment by sherlcok314159** [comment]
Thanks for your immediate reply. But this raises an interesting question, "better performance may be not very positively related to the balanced FFN layer grad norm distribution". The HybridNorm* achieved the best performance. Right?

**Comment by BryceZhuo** [comment]
That’s a great question, and we’d like to offer some preliminary thoughts on it.

While HybridNorm* does achieve the best performance in our experiments, it also involves special handling in the first layer, which results in a slightly different (and less balanced) gradient norm distribution—particularly in the early layers. This observation suggests that a perfectly balanced FFN gradient norm distribution does not necessarily correlate with better performance.

Instead, our findings indicate that the relationship between gradient norm distribution and model performance is more nuanced. The modifications in HybridNorm* may introduce beneficial inductive biases or subtly enhance the optimization dynamics. As noted in [1,2], the initial and final layers of a Transformer are functionally distinct, whereas the middle layers tend to be more interchangeable. This might imply that what’s important is not achieving balance across all layers uniformly, but rather maintaining an appropriate balance within each group of layers.

[1] Skean, et al. [Does Representation Matter? Exploring Intermediate Layers in Large Language Models](https://arxiv.org/pdf/2412.09563).
[2] Sun, et al. [Transformer Layers as Painters](https://arxiv.org/pdf/2407.09298).
